### ISSUE TITLE
Add support for PEP 695

### ISSUE THREAD

**Issue by charliermarsh** [astral-sh/ruff#5062]
## Summary

[PEP 695](https://peps.python.org/pep-0695/) adds new syntax for type parameters and generics. A few examples:

```py
class ClassA[T: str]:
    def method1(self) -> T:
        ...

type ListOrSet[T] = list[T] | set[T]
```

There are a few pieces of work involved in this:

- Adding these elements to the AST.
- Adding these elements to the parser (especially the `type` soft keyword).
- Adding these elements to the visitor, and to any rules that do custom visiting.
- Adding these elements to our semantic model (need to look into how scoping and resolution works).

This will be Ruff's first time undergoing a significant grammar and syntax change (apart from adding structural pattern matching), so it's a great opportunity to set the project's standards for keeping up with Python's evolution.

Specifically, the final Python 3.12 release is expected on 2023-10-02. We should aim to support PEP 695 by that date _at the latest_.

**Comment by charliermarsh** [comment]
\cc @zanieb who's working on this :)

**Comment by zanieb** [comment]
Progress on the Rust Python parser can be tracked at https://github.com/RustPython/Parser/issues/82

**Comment by zanieb** [comment]
Additional resource; may be useful for refinements to scoping https://jellezijlstra.github.io/pep695
